app/src/main/java/org/mozilla/tv/firefox/pinnedtile/PinnedTileRepo.kt

```diff
@@ -0,0 +1,131 @@
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+package org.mozilla.tv.firefox.pinnedtile
+
+import android.app.Application
+import android.arch.lifecycle.MutableLiveData
+import android.content.Context
+import android.content.SharedPreferences
+import android.graphics.Bitmap
+import android.graphics.BitmapFactory
+import android.support.annotation.AnyThread
+import android.support.annotation.UiThread
+import org.json.JSONArray
+import java.util.UUID
+
+private const val BUNDLED_SITES_ID_BLACKLIST = "blacklist"
+private const val CUSTOM_SITES_LIST = "customSitesList"
+private const val PREF_HOME_TILES = "homeTiles"
+private const val BUNDLED_HOME_TILES_DIR = "bundled"
+private const val HOME_TILES_JSON_PATH = "$BUNDLED_HOME_TILES_DIR/bundled_tiles.json"
+
+/**
+ * TODO
+ * Some methods require applicationContext in order to access /assets/
+ */
+class PinnedTileRepo(private val applicationContext: Application) {
+    private val _pinnedTiles = MutableLiveData<LinkedHashMap<String, PinnedTile>>() // FIXME: URI? String?
+    val tileCount get() = _pinnedTiles.value?.size
+
+    // TODO:  applicationContext.getSharedPreferences(PREF_HOME_TILES, Context.MODE_PRIVATE)
+    private var _sharedPreferences: SharedPreferences = applicationContext.getSharedPreferences(PREF_HOME_TILES, Context.MODE_PRIVATE)
+
+    fun loadTilesCache(): MutableLiveData<LinkedHashMap<String, PinnedTile>> {
+        val pinnedTiles = linkedMapOf<String, PinnedTile>().apply {
+            putAll(loadBundledTilesCache())
+            putAll(loadCustomTilesCache())
+        }
+
+        _pinnedTiles.value = pinnedTiles
+
+        return _pinnedTiles
+    }
+
+    fun addPinnedTile(url: String, screenshot: Bitmap?) {
+        val newPinnedTile = CustomPinnedTile(url, "custom", UUID.randomUUID()) // TODO: titles
+        _pinnedTiles.value?.put(url, newPinnedTile)
```

Comment:
Maybe we should:
```kotlin
fun addPinnedTile(...) {
    persistCustomTiles()
    pinnedTiles.value = loadCachedTiles()
}
```

It's less performant but it hides bugs in the duplicated logic of mutating the pinnedTiles in place. Since we don't call this often and our lists are small, it seems acceptable.

Then again, there's already a hiccup when pinning sites due to the screenshot so maybe we don't want to make that worse.

Reply:
I will create an issue about this so we can address it better (optimization)